Check whether data imported from Whitehall is accurate with the Publishing API

leenagupte opened the pull request:
Trello: https://trello.com/c/JHMoTkfA
Based on: https://github.com/alphagov/content-publisher/pull/1577
Depends on: https://github.com/alphagov/gds-api-adapters/pull/975

## What's changed
Creates a service that checks whether the data we are using to replace Whitehall data during the migration is accurate with the Publishing API. This service acts as the final buffer as to whether to complete a migration or abort.

## Why?
To give confidence when doing a bulk republishing as part of migration. 

## Testing in integration
### Matches
```
document = Document.find_by_param("e882a7b1-8605-45e7-9700-7b0fe8e068d0:en")
=> #<Document id: 671, content_id: "e882a7b1-8605-45e7-9700-7b0fe8e068d0", locale: "en", created_at: "2019-08-02 07:38:33", updated_at: "2019-12-30 14:58:44", created_by_id: 139, first_published_at: nil, imported_from: "whitehall">
irb(main):002:0> WhitehallImporter::IntegrityChecker.new(document.current_edition).valid?
=> true
irb(main):004:0> WhitehallImporter::IntegrityChecker.new(document.current_edition).problems
=> []
```

### Doesn't match
```
irb(main):008:0> document = Document.find_by_param("1dd15d42-e6b5-4fab-8492-350eca33cf11:en")
=> #<Document id: 667, content_id: "1dd15d42-e6b5-4fab-8492-350eca33cf11", locale: "en", created_at: "2019-01-10 14:44:21", updated_at: "2019-01-23 17:38:10", created_by_id: 130, first_published_at: nil, imported_from: "whitehall">
irb(main):009:0> WhitehallImporter::IntegrityChecker.new(document.live_edition).valid?
=> false
irb(main):010:0> WhitehallImporter::IntegrityChecker.new(document.live_edition).problems
=> ["image caption doesn't match"]

proposed = PreviewService::Payload.new(edition, republish: true).payload
proposed["details"]["image"]
=> {"high_resolution_url"=>nil, "url"=>nil, "alt_text"=>"Man being interviewed in front of a fire engine", "caption"=>"", "credit"=>nil}
```

erino commented on lib/whitehall_importer/integrity_checker.rb:
```diff
@@ -0,0 +1,98 @@
+      problems += image_problems
+      problems += organisation_problems
+
+      problems
```
What about 

```ruby
[].concat(content_problems, image_problems, organisation_problems)
```

leenagupte replied:
I didn't realise `concat` is a lot faster.

leenagupte replied:
I could just do `content_problems + image_problems + organisation_problems`

erino replied:
Even better

kevindew commented on Gemfile:
```diff
@@ -24,6 +24,7 @@ gem "pg", "~> 1"
 gem "plek", "~> 3"
 gem "rinku", "~> 2"
 gem "rubyzip", "~> 2", require: "zip"
+gem "sanitize", "~> 5.1.0"
```
Can we go with `~> 5` to be consistent with our other pins?

kevindew commented on lib/whitehall_importer/import.rb:
```diff
@@ -73,5 +74,16 @@ def first_publish_date
+    end
+
+    def integrity_check_edition(edition)
+      integrity_checker = WhitehallImporter::IntegrityChecker.new(edition)
```
You shouldn't need this to be prefixed with "WhitehallImporter::"

leenagupte replied:
Oops! Will fix.

kevindew commented on lib/whitehall_importer/import.rb:
```diff
@@ -73,5 +74,16 @@ def first_publish_date
+      integrity_check_edition(document.current_edition) if document.current_edition != document.live_edition
+    end
+
+    def integrity_check_edition(edition)
```
These two methods may feel more consistent with the methods with the rest of this file if they start with the verb. For example `check_document_integrity` and `check_edition_integrity`

kevindew commented on lib/whitehall_importer/integrity_checker.rb:
```diff
@@ -0,0 +1,93 @@
+
+      problems = []
+
+      %w(alt_text caption).each do |attribute|
```
You can use an each_with_object or an inject to avoid the need to create and return the problems array.

kevindew commented on spec/factories/whitehall_export/edition_factory.rb:
```diff
@@ -52,5 +52,28 @@
+    trait :published do
+      transient do
+        previous_state { "submitted" }
+      end
```
Do we make use of this transient attribute? I have a feeling it's a c+p from scheduled where we care more about it

leenagupte replied:
You caught me! I'll get rid of it.

kevindew commented on spec/factories/whitehall_export/edition_factory.rb:
```diff
@@ -52,5 +52,28 @@
+                state: "published",
+                created_at: Time.current.tomorrow.rfc3339),
+        ]
+      end
```
I see there's other scheduled bits in here. I wonder if it'd work the same for us were we to do `build(:whitehall_export_edition, state: "published")` and avoid needing the trait?

leenagupte replied:
I tried that, but it doesn't work because you also need a revision history entry for the "published" state, so I thought it was nicer to create a trait instead.

kevindew replied:
No problem 👍

kevindew commented on spec/lib/whitehall_importer/import_spec.rb:
```diff
@@ -107,5 +113,28 @@
+                    ))
+
+      expect { described_class.call(build(:whitehall_export_document)) }
+        .to raise_error(WhitehallImporter::AbortImportError)
```
Should we also assert that this includes the message since we typed it out above?

kevindew commented on spec/lib/whitehall_importer/integrity_checker_spec.rb:
```diff
@@ -0,0 +1,144 @@
+          organisations: [SecureRandom.uuid],
+        },
+      )
+    end
```
Can we use build(:edition) rather than create so we don't need to hit the database?

kevindew commented on spec/lib/whitehall_importer/integrity_checker_spec.rb:
```diff
@@ -0,0 +1,144 @@
+  end
+
+  describe "#problems" do
+    let(:edition) { create(:edition) }
```
similar point on build rather than create

leenagupte commented:
@kevindew Thanks for your review. I've addressed your comments and fixed up. The only one I wasn't sure about, and have left, is the one about the number of tests, but we spoke about that IRL.

kevindew commented on spec/lib/tasks/import_spec.rb:
```diff
@@ -28,12 +28,14 @@
       Rake::Task["import:whitehall_document"].reenable
       allow(ResyncService).to receive(:call)
       allow(WhitehallImporter::ClearLinksetLinks).to receive(:call)
+      allow(WhitehallImporter::Import).to receive(:call).and_return(create(:document, :with_current_edition))
```
I think we've got an issue here where this is also testing a level above the unit of code. 

I've put together a quick proposal for how we should unit test this task: https://github.com/alphagov/content-publisher/commit/a99eed212c58ac7adf7e72727934dbdffe3cfa7e

leenagupte replied:
Oh! You want to remove the "creates a document" test? Well that certainly makes the rake task easier to test. I assume it's because the functionality and whether a document is created should be tested by the `WhitehallImporter` tests?

kevindew replied:
Yeah we don't actually create a document in the rake task so it seems an odd thing to assert on, I guess we either used to do that or we were testing this more like an integration test.

kevindew commented on lib/whitehall_importer/integrity_checker.rb:
```diff
@@ -0,0 +1,97 @@
+    def primary_publishing_organisation_matches?
+      primary_publishing_organisation =
+        publishing_api_content.dig("links", "primary_publishing_organisation") ||
+        publishing_api_links.dig("links", "primary_publishing_organisation")
```
Do we want to capture this in a method?

we could have: 

```
def publishing_api_link(link_type)
  publishing_api_content.dig("links", link_type) ||
    publishing_api_links.dig("links", link_type)
end
```

leenagupte replied:
Ha! That's what I had originally and took out! I can add it back in.

kevindew commented on spec/lib/whitehall_importer/integrity_checker_spec.rb:
```diff
@@ -0,0 +1,168 @@
+
+      integrity_check = WhitehallImporter::IntegrityChecker.new(edition)
+      expect(integrity_check.valid?).to be true
+    end
```
👍

leenagupte commented:
@kevindew I've addressed the rest of your comments and fixed up. Cheers!
